internal/service/resource_pool_object_mapper.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2023 Red Hat Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
+compliance with the License. You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under the License is
+distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+implied. See the License for the specific language governing permissions and limitations under the
+License.
+*/
+
+package service
+
+import (
+	"context"
+	"errors"
+	"log/slog"
+
+	"github.com/openshift-kni/oran-o2ims/internal/data"
+	"github.com/openshift-kni/oran-o2ims/internal/streaming"
+)
+
+//go:generate mockgen -source=resource_pool_object_mapper.go -package=service -destination=resource_pool_object_mapper_mock.go
+type ResourcePoolObjectMapper interface {
+	MapItem(ctx context.Context, from data.Object) (to data.Object, err error)
+}
+
+type resourcePoolObjectMapper struct {
+	cloudID   string
+	resources data.Stream
+}
+
+func NewResourcePoolObjectMapper(cloudID string, resources data.Stream) ResourcePoolObjectMapper {
+	return &resourcePoolObjectMapper{
+		cloudID:   cloudID,
+		resources: resources,
+	}
+}
+
+// MapItem is mapping Cluster objects fetched from search API into O2 ResourcePool objects
+func (r *resourcePoolObjectMapper) MapItem(ctx context.Context, from data.Object) (to data.Object, err error) {
+	resourcePoolID, err := data.GetString(from, "cluster")
+	if err != nil {
+		return
+	}
+
+	name, err := data.GetString(from, "name")
+	if err != nil {
+		return
+	}
+
+	labels, err := data.GetString(from, "label")
+	if err != nil {
+		return
+	}
+	labelsMap := data.GetLabelsMap(labels)
+
+	resources := r.getResourceArray(ctx, r.resources, resourcePoolID)
+
+	to = data.Object{
+		"resourcePoolID": resourcePoolID,
+		"name":           name,
+		"oCloudID":       r.cloudID,
+		"extensions":     labelsMap,
+		"resources":      resources,
+		// TODO: no direct mapping to a property in Cluster object
+		"description":      "",
+		"location":         "",
+		"globalLocationID": "",
+	}
+	return
+}
+
+// Converts a specified data stream of O2 Resources (Nodes) to an array.
+func (r *resourcePoolObjectMapper) getResourceArray(ctx context.Context, items data.Stream, resourcePoolID string) (result []map[string]any) {
```

Comment:
In general we should try to avoid things that return arrays, because they negate the value of using streams. But if we do then we can combine the `Collect` and `Select` methods. For example:

```go
selector := func(ctx context.Context, item data.Object) (result bool, err error) {
  result = item["resourcePoolID"] == resourcePoolID
  return
}
result = streams.Collect(streams.Select(items, selector))
```